src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs

```diff
@@ -142,37 +155,61 @@ private bool IsTopLevelWildcard(string host)
         }
 
         // This does not duplicate format validations that are expected to be performed by the host.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private bool CheckHost(HttpContext context, IList<StringSegment> allowedHosts)
         {
-            var host = new StringSegment(context.Request.Headers[HeaderNames.Host].ToString()).Trim();
+            var host = context.Request.Headers[HeaderNames.Host].ToString();
 
-            if (StringSegment.IsNullOrEmpty(host))
+            if (0u >= (uint)host.Length)
             {
-                // Http/1.0 does not require the host header.
-                // Http/1.1 requires the header but the value may be empty.
-                if (!_options.AllowEmptyHosts)
-                {
-                    _logger.LogInformation("{Protocol} request rejected due to missing or empty host header.", context.Request.Protocol);
-                    return false;
-                }
-                _logger.LogDebug("{Protocol} request allowed with missing or empty host header.", context.Request.Protocol);
-                return true;
+                return IsEmptyHostAllowed(context);
+            }
+
+            // Host header should not require trimming
+            // rfc7230  3.2.4.  Field Parsing https://tools.ietf.org/html/rfc7230#page-25
+            // Optional whitespace occurring before the first non-whitespace octet of the field value or after the last
+            // non-whitespace octet of the field value ought to be excluded by
+            // parsers when extracting the field value from a header field
+            if (char.IsWhiteSpace(host[0]) || char.IsWhiteSpace(host[host.Length - 1]))
```

Comment:
At this point your enforcing a header parsing rule. If there's an issue here it's with the underlying request parsing, not with the client's request. You can remove this whitespace check.

Reply:
Dropped